### ISSUE TITLE
My metrics aren't improving with more epochs

### ISSUE THREAD

**Issue by JesusSilvaUtrera** [ultralytics/yolov5#10082]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I'm trying to train on a custom dataset to detect terminals on industrial pieces for a school project, but I'm not getting good results. In fact, I'm not getting results at all, because I only get mAP=0 or mAP~=0. Looking at the train_batch images, it looks like the training is getting the labels right, but the predictions are horrible, looking at the val_batch images I see that the model doesn't predict anything. 

Train_batch:
![train_batch1](https://user-images.githubusercontent.com/79662866/200602508-0582732a-d783-49c8-9b32-922485f1a2ac.jpg)

Val_batch labels are good:
![val_batch0_labels](https://user-images.githubusercontent.com/79662866/200602700-19340e34-ef20-4d43-a3a4-3abed4220d8d.jpg)

But no predictions are returned:
![val_batch0_pred](https://user-images.githubusercontent.com/79662866/200602775-ff49c244-ab6d-458c-8f6d-78c13e49f768.jpg)

I started with only 10 images and 5 epochs to see if it worked (I know they are too few, but it was just to test that it worked and it should return at least something, didn't it?), and as I didn't receive any error I assume that the code is receiving all needed for the training. I also tried to lower conf_thresh and iou_thresh to 0.01 in the validation to see if it improves but it does the same. Am I doing something wrong? Or is it just that I need more epochs and images to train? I wanted to train with just 100 images or so to make the training not too long or computationally expensive.

**Comment by JesusSilvaUtrera** [comment]
I forgot to add the results.png file to see the actual behaviour of the model:
![results](https://user-images.githubusercontent.com/79662866/200604243-5616fbd2-f66c-4ec0-951e-77bba2de4efd.png)

**Comment by ExtReMLapin** [comment]
There is only 4 epochs ? Try with at least 400

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by JesusSilvaUtrera** [comment]
Okay, I will try with more images and epochs then and see what happens

**Comment by JesusSilvaUtrera** [comment]
@ExtReMLapin @glenn-jocher I have tried with the coco128 dataset to see if it goes well, and it has good mAP but the metrics doesn't change with the passage of the epochs, it starts in mAP@.5 = 0.704 and it remains like that for all epochs, it seems the network is not updating the weights maybe? Same happens with other metrics, except for the train loss (not improving so much though). Here are the results for 30 epochs (it's still few epochs, but I'm working on VS Code to debug any issues and I don't have a GPU):
![results](https://user-images.githubusercontent.com/79662866/200894239-e73938ac-3aee-46cd-8cad-3fb2787bfb64.png)

**Comment by JesusSilvaUtrera** [comment]
@ExtReMLapin @glenn-jocher have you experienced something similar before? It's like all layers are frozen and it doesn't train at all

**Comment by glenn-jocher** [comment]
@JesusSilvaUtrera there's something wrong your model or dataset, or YOLOv5 code. I'd go to our colab notebook to see what training should look like with a verified environment, and then work from there:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb?hl=en

Just run Setup cell and then go to Train section and run train.py cell.

**Comment by JesusSilvaUtrera** [comment]
@glenn-jocher thank you for your quick response, I will try to discover what is happening

**Comment by ExtReMLapin** [comment]
As glenn said, precision  and recall being static at the same value for multiple epoch is more than suspicious.
Are you sure you correctly split your dataset into train/test/valid and there is no intersection between the 3 ? (they don't share any identical data)

**Comment by glenn-jocher** [comment]
@ExtReMLapin great points! Overfitting could be a problem here as well, but data leakage due to improper dataset splitting is also suspicious and should be checked thoroughly. You should definitely ensure there's no data contamination between your train/test/val sets.

Our [documentation](https://docs.ultralytics.com/yolov5/training) outlines correct data preparation and usage, including dataset splitting. Make sure to follow those instructions meticulously. Additionally, sharing the commandline used to run the training, even without dataset specifics, could be invaluable in diagnosing any potential misconfigurations.
